development/webpack/utils/plugins/ManifestPlugin/index.ts

```diff
@@ -0,0 +1,347 @@
+        browserManifest = {
+          ...browserManifest,
+          ...JSON.parse(readFileSync(browserManifestPath, 'utf8')),
+        };
+      } catch {
+        // ignore if the file doesn't exist, as some browsers might not need overrides
+      }
+
+      // merge provided `web_accessible_resources`
+      const resources = this.options.web_accessible_resources;
+      if (resources && resources.length > 0) {
+        if (browserManifest.manifest_version === 3) {
+          browserManifest.web_accessible_resources =
+            browserManifest.web_accessible_resources || [];
+          const war = browserManifest.web_accessible_resources.find(
+            (resource) => resource.matches.includes('<all_urls>'),
+          );
+          if (war) {
+            // merge the resources into the existing <all_urls> resource, ensure uniqueness using `Set`
+            war.resources = [...new Set([...war.resources, ...resources])];
+          } else {
+            // add a new <all_urls> resource
+            browserManifest.web_accessible_resources.push({
+              matches: ['<all_urls>'],
+              resources: [...resources],
+            });
+          }
+        } else {
+          browserManifest.web_accessible_resources = [
+            ...(browserManifest.web_accessible_resources || []),
+            ...resources,
+          ];
+        }
+      }
+
+      // allow the user to transform the manifest. Use a copy of the manifest
+      // so modifications for one browser don't affect other browsers.
+      browserManifest =
+        this.options.transform?.(
+          JSON.parse(JSON.stringify(browserManifest)),
+          browser,
+        ) || browserManifest;
+
+      // Add the manifest file to the assets
+      const source = new RawSource(JSON.stringify(browserManifest, null, 2));
+      this.manifests.set(browser, source);
+    });
+  }
+
+  private hookIntoPipelines(compilation: Compilation): void {
+    // prepare manifests early so we can catch errors early instead of waiting
+    // until the end of the compilation.
+    this.prepareManifests(compilation);
+
+    // TODO: MV3 needs to be handled differently. Specifically, it needs to
+    // load the files it needs via a function call to `importScripts`, plus some
+    // other shenanigans.
+
+    // hook into the processAssets hook to move/zip assets
+    const tapOptions = {
+      name: NAME,
+      stage: Infinity,
+    };
+    if (this.options.zip) {
+      const options = this.options as ManifestPluginOptions<true>;
+      compilation.hooks.processAssets.tapPromise(
+        tapOptions,
+        async (assets: Assets) => {
+          await this.zipAssets(compilation, assets, options);
+          this.moveAssets(
+            compilation,
+            assets,
+            this.options as ManifestPluginOptions<false>,
+          );
+        },
+      );
+    } else {
+      const options = this.options as ManifestPluginOptions<false>;
```

Comment:
I couldn't make TypeScript figure out these casts without a bunch of extra lines or needlessly making extra copies of the `options` object, so I've just casted them here. It's fine. Look away!